### PR TITLE
JIT: refactor how opaque VNs are represented

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Use a unary function to capture opaque VN loop dependence, rather than factoring
that out as part of the owning chunk. As a result the VN chunk data no longer
has a per-loop component (no dependence on MAX_LOOP_NUM).

This gives us the flexibility to address #54118 by doing something similar for
`MapUpdate` VNs.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@Briansull PTAL   
cc @dotnet/jit-contrib 

No diffs (per SPMI).

I can add a second commit with the remainder of the fix for #54118 here, but would prefer to merge this as a preparatory no-diff change.

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -6582,21 +6587,6 @@ bool Compiler::optVNIsLoopInvariant(ValueNum vn, unsigned lnum, VNToBoolMap* loo
             }
         }
     }
-    else
```
Comment (briansull) [inline src/coreclr/jit/optimizer.cpp:6585]: You are removing this block of code.

Was this used prior to your changes?

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -454,7 +454,7 @@ ValueNumStore::ValueNumStore(Compiler* comp, CompAllocator alloc)
     // We have no current allocation chunks.
     for (unsigned i = 0; i < TYP_COUNT; i++)
     {
-        for (unsigned j = CEA_None; j <= CEA_Count + BasicBlock::MAX_LOOP_NUM; j++)
+        for (unsigned j = CEA_Const; j <= CEA_Count; j++)
```
Comment (briansull) [inline src/coreclr/jit/valuenum.cpp:457]: Is this change a bug-fix, as we weren't fully initializing things?

Comment (briansull) [reply src/coreclr/jit/valuenum.cpp:457]: I see now that you are just removing the MAX_LOOP_NUM stuff.
You did change the start of this loop from CEA_None to CSE_Const

File: src/coreclr/jit/optimizer.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:6585]: The then clause now handles that bit  of logic (via the `VNF_MemOpaque` addition).

File: src/coreclr/jit/valuenum.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:457]: Yep, we no longer need the `CEA_None` chunks.

File: src/coreclr/jit/valuenum.cpp [hunk 3]
```diff
@@ -3675,12 +3654,14 @@ ValueNum ValueNumStore::VNForExpr(BasicBlock* block, var_types typ)
...
-    unsigned offsetWithinChunk = c->AllocVN();
-    ValueNum result            = c->m_baseVN + offsetWithinChunk;
-    return result;
+    // VNForFunc(typ, func, vn) but bypasses looking in the cache
+    //
+    VNDefFunc1Arg fstruct(VNF_MemOpaque, loopNum);
+    Chunk*        c                                                = GetAllocChunk(typ, CEA_Func1);
+    unsigned      offsetWithinChunk                                = c->AllocVN();
+    ValueNum      resultVN                                         = c->m_baseVN + offsetWithinChunk;
+    reinterpret_cast<VNDefFunc1Arg*>(c->m_defs)[offsetWithinChunk] = fstruct;
```
Comment (briansull) [inline src/coreclr/jit/valuenum.cpp:3663]: This assignment looks odd and is hard to parse.
If it can't be written as a typical assignment, then add a comment with an explanation.

Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:3663]: You mean the last line there? That is what all of the `VNForFunc` overloads do -- I just "inlined" the unary case here since for these opaque VNs we don't memoize.

Comment (briansull) [reply src/coreclr/jit/valuenum.cpp:3663]: I see it written here using two lines:

```
ValueNum ValueNumStore::VnForConst(T cnsVal, NumMap* numMap, var_types varType)

        T* chunkDefs                 = reinterpret_cast<T*>(chunk->m_defs);
        chunkDefs[offsetWithinChunk] = cnsVal;
```

File: src/coreclr/jit/valuenum.cpp [hunk 4]
```diff
@@ -4343,27 +4324,30 @@ var_types ValueNumStore::TypeOfVN(ValueNum vn)
...
+//    Returns BasicBlock::MAX_LOOP_NUM if this VN is not a memory value number.
+//
 BasicBlock::loopNumber ValueNumStore::LoopOfVN(ValueNum vn)
 {
-    if (vn == NoVN)
+    if (IsVNFunc(vn))
     {
-        return BasicBlock::MAX_LOOP_NUM;
+        VNFuncApp funcApp;
+        GetVNFunc(vn, &funcApp);
```
Comment (jakobbotsch) [inline src/coreclr/jit/valuenum.cpp:4342]: I can't add a suggestion since there's a deleted line in the diff, but `GetVNFunc` returns a bool so you can combine it with the check.

File: src/coreclr/jit/valuenum.cpp (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:3663]: Ok... might as well change this everywhere it appears.

File: src/coreclr/jit/valuenum.cpp (hunk 4 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:4342]: Thanks. Will update.

**Review by jakobbotsch — APPROVED** [review]

**Review by briansull — APPROVED** [review]
Looks Good
